### PR TITLE
Add network types crate and versioned crate

### PR DESCRIPTION
**Pull request by willemolding**
**Summary of changes**

I've split issue #138 into a number of PRs for ease of review. This one only adds things.

Changes introduced in this pull request:
- Adds the `network-types` crates which collates all of the types used to generate the serde serialization code
- Copies some tests from the test-serialization crate to become the tests for the new crate. Propose removing the tests from the other crate in a future PR
- Adds the `versioned` crate  which supports the `network-types` crate and adds a more explicit method for versioning the types that does not require a proc macro. Types now carry their version info with them. This means if we do need to update a type to a later version the type-checker will be able to guide us through the update. This replaces the `wire-type` crate which should also be removed in a future PR.

### CODE REVIEW

**Review by polarbeargo — APPROVED** [review]
An astonished, amazing, and elegant work on resolving our protocol version problem 🚀

File: protocol/network-types/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+edition = "2021"
+license = "Apache-2.0"
+
+[dependencies]
+versioned = {path = "../versioned"}
+serde = {version = "1", features = ["derive"]}
+mina-crypto = {path = "../../crypto"} # TODO: Try and remove this dep
+
+[dev-dependencies]
+wasm-bindgen-test = "0.3"
```
Comment (hanabi1224) [inline protocol/network-types/Cargo.toml:15]: Where is `wasm-bindgen-test` used? Plz note that wasm tests in new crates need extra setup in CI workflow yaml

File: protocol/network-types/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+[package]
+authors = ["ChainSafe Systems <info@chainsafe.io>"]
+name = "mina-network-types"
```
Comment (hanabi1224) [inline protocol/network-types/Cargo.toml:3]: The serialized bytes can also be used for storage, maybe `serializable-types` is better than `network-types`?

**Review by hanabi1224 — APPROVED** [review]
Nice work! Only some minor comments on naming

**Pull request merged**